[grpc/protobuf] Update grpc to 1.65.5 and update protobuf to 5.26.1

Tradias opened the pull request:
Fixes
https://github.com/microsoft/vcpkg/issues/37440
https://github.com/microsoft/vcpkg/issues/29697
https://github.com/microsoft/vcpkg/issues/38316
https://github.com/microsoft/vcpkg/issues/37729
https://github.com/microsoft/vcpkg/issues/30583
https://github.com/microsoft/vcpkg/issues/40803

Upstream issues:
https://github.com/Ultimaker/libArcus/issues/156
https://github.com/apache/brpc/issues/2757
https://github.com/sogou/srpc/issues/406

Update to latest protobuf isn't possible because gRPC latest does not support it and it is difficult to patch (unstable upb API being the primary reason).

Some ports had to be patched to support protobuf v5.

upb now has (hopefully) proper stage0 build to generate `descriptor.upb.h/c` files. Seeing that the supposedly tested-for-staleness files here: https://github.com/protocolbuffers/protobuf/tree/v25.1/upb/cmake/google/protobuf are actually stale.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Tradias committed:
grpc: Update to 1.65.0

Tradias committed:
upb: Fix using host tools

Tradias commented:
@MonicaLiu0311 ready for review.

MonicaLiu0311 commented:
grpc & protobuf & upb
-
All features are tested successfully in the following triplet:
```
x86-windows
x64-windows
x64-windows-static
```


openvino
-
Features `onnx,paddle` are tested successfully in the following triplet:
```
x64-windows
x64-windows-static
```

MonicaLiu0311 commented:
@Tradias Please resolve the conflict.

Tradias committed:
protobuf: Patch/update arcus, brpc, openvino. Fix grpc otel pkgconfig. Add proper stage0 to upb

JavierMatosD commented on ports/brpc/protobuf.patch:
```diff
@@ -1,5 +1,262 @@
+diff --git a/src/brpc/esp_message.cpp b/src/brpc/esp_message.cpp
```
Can you explain where this patch is coming from? What problem does it solve and why is it needed to update `grpc` and `protobuf`? In general, this patch looks upstream-able.

Tradias replied:
In addresses the breaking changes caused by the protobuf update from v4 to v5. It can certainly be upstreamed.

JavierMatosD commented on ports/arcus/0001-fix-protobuf-deprecated.patch:
```diff
@@ -1,3 +1,16 @@
+     ErrorCollector() : _error_count(0) { }
+ 
+-    void AddError(const std::string& filename, int line, int column, const std::string& message) override
++    void RecordError(absl::string_view filename, int line, int column, absl::string_view message) override
```
Please explain why this patch is needed. This looks like something that is upstream-able.

Tradias replied:
It addresses the breaking changes caused by the protobuf update from v4 to v5. It can certainly be upstreamed.

Benjamin-Miller1 replied:
@JavierMatosD It's been a long time. Have you noticed?

BillyONeal replied:
Please submit it upstream. (Not a blocker for this PR)

@Benjamin-Miller1 This PR has been in a big state of flux for a long time :)

Tradias replied:
I leave that to the arcus maintainers. That's what https://github.com/Ultimaker/libArcus/issues/156 is for

JavierMatosD reviewed (dismissed):
Partial Review:

This PR is claiming to update `grpc` and `protobuf`, but I am also seeing updates for:

- `utf8-range`
- `upb`
- `srpc`
- `openvino`
- `brpc`
- `arcus` 

Why are these changes needed for the version update? If they are not, please split apart into their own respective PRs. I'm not saying that the changes here wouldn't be accepted just that they should be reviewed on their own. 

> Update to latest protobuf isn't possible because gRPC latest does not support it and it is difficult to patch (unstable upb API being the primary reason).

Can you explain why `grpc` doesn't support latest `protobuf`? i.e., did `grpc` drop support for `protobuf` or is this an upstream bug? In general, we don't accept patches that add functionality that isn't present in upstream. That is, if `grpc` doesn't support latest `protobuf` _and that is by design_ then we will not add the functionality through vcpkg. We typically only accept version updates to latest barring unique circumstances so this needs a bit more explanation.  

In general, there are a lot of modifications to patches here. I need a bit of context as to why you are making those changes and whether or not they have been reported to upstream. Overall, these changes look fine, but I want to make sure we aren't creating more maintenance burden here by adding support for stuff that upstream by design does not support.

Tradias commented:
@JavierMatosD Thank you for looking into this pr (again). The story is very similar to the previous one https://github.com/microsoft/vcpkg/pull/35781. We are trying to force a protobuf version onto libraries that don't support it (yet) or are unmaintained. I haven't created any reports upstream and honestly don't feel like doing it, so feel free to do so.

`protobuf`, `utf8-range` and `upb` all come from the same repository. I think it would be best to avoid updating one without the others. That also means that we cannot update to latest protobuf (which is support by gRPC) because the latest upb is not.

Tradias committed:
grpc: Update to 1.65.2

Tradias committed:
upb: Integrate upb de-vendoring changes

Tradias committed:
protobuf: Adjust fix-arm64-msvc patch to v5.26.1

Tradias committed:
grpc: Update to 1.65.4

Tradias committed:
brpc: Fix protobuf patch for version 1.10.0

pedrolamarao commented:
Thank you for your work on upgrading gRPC!
Is there anything the community can do to help this PR move forward?

Tradias commented:
Upstream "support protobuf v5" issues:


https://github.com/Ultimaker/libArcus/issues/156
https://github.com/apache/brpc/issues/2757
https://github.com/sogou/srpc/issues/406

AlangabSistemi commented:
When will this change be accepted? I need uds in windws (grpc>=1.63)

pedrolamarao commented:
I have started to experiment with this branch to prepare my project for this upgrade.

It seems that I have hit a known problem with protobuf.

```
P:\dev\keys-native\build\vcpkg_installed\x64-windows-static-md\include\google\protobuf\map_field.h(682,66): error C2370
: 'google::protobuf::internal::MapField<Derived,Key,T,kKeyFieldType_,kValueFieldType_>::kVTable': redefinition; differe
nt storage class [P:\dev\keys-native\build\protobuf\keys-protobuf.vcxproj]
P:\dev\keys-native\build\vcpkg_installed\x64-windows-static-md\include\google\protobuf\map_field.h(682,66): error C2370
: 'google::protobuf::internal::MapField<Derived,Key,T,kKeyFieldType_,kValueFieldType_>::kVTable': redefinition; differe
nt storage class [P:\dev\keys-native\build\protobuf\keys-protobuf.vcxproj]
  (compiling source file 'src/main/proto/keys.pb.cc')
      P:\dev\keys-native\build\vcpkg_installed\x64-windows-static-md\include\google\protobuf\map_field.h(671,37):
      see declaration of 'google::protobuf::internal::MapField<Derived,Key,T,kKeyFieldType_,kValueFieldType_>::kVTable'

  (compiling source file 'src/main/proto/keys.grpc.pb.cc')
      P:\dev\keys-native\build\vcpkg_installed\x64-windows-static-md\include\google\protobuf\map_field.h(671,37):
      see declaration of 'google::protobuf::internal::MapField<Derived,Key,T,kKeyFieldType_,kValueFieldType_>::kVTable'
```

This seems the same symptom as https://github.com/microsoft/vcpkg/issues/39459

Tradias commented:
@pedrolamarao You describe an issue that exists in the protobuf version on master. In the linked vcpkg issue you will find the protobuf commit that fixed it and if you check it out you can see that the fix made it into the protobuf version that this pull request introduces https://github.com/protocolbuffers/protobuf/commit/f78f9c51fa2470070e5d4b49649800971c789224. I also just compiled a small program with vcpkg master branch (failed as expected) and then with this branch, succeeded.

In other words, it looks like you did something wrong when trying to test this branch and instead used current master branch. Possible causes:

* Forgot to remove `builtin-baseline` from your vcpkg.json file
* VCPKG_ROOT env value is pointing at a different clone of vcpkg instead of the one where you checked out this branch

pedrolamarao commented:
@Tradias thanks, that sounds likely, sorry for the noise.

korDen commented:
What needs to be done to move this forward? @JavierMatosD are you able to do a second pass?

MonicaLiu0311 commented:
@BillyONeal Please take a look at this PR.

FAKERINHEART commented:
hey, buddies when will it be merge?

JavierMatosD commented:
@BillyONeal, @vicroms, @data-queue discussed this today.

@BillyONeal - "`protobuf` and `upb` are owned by the same people, we should be picking versions that are intrinsically compatible with each other. If that means we can't update `grpc`, so be it. We are not `upb`, `grpc`, or `protobuf` maintainers and we are not qualified to review that. We have to pick some form of versioning/packaging story that preserves that and if it means we can't update that then we can't update that. We would need to file issues with upstream and get them to implement the fix."

@ras0219-msft - "It is a reimplementation of protobuf that is small, so it has independent implementation from protobuf but it is c, small, and unstable. It sounds like `upb` is not suitable for consumption. We should declare that `upb` only exists as a dependency of `grpc` and treated as such. Maybe the path forward is `protobuf` and `upb` become decoupled and `upb` and `grpc` become coupled."

@BillyONeal - "Maybe `upb` should be delisted"

@ras0219-msft - "Why was `upb` added in the first place?"

https://github.com/microsoft/vcpkg/issues/8460#issue-502997609 -> implies it is needed for `grpc`

@ras0219-msft - "`grpc` has their own vendor copies of all these dependencies. So presumably, they have a set of dependencies that works for them. We could just say that `grpc` owns this whole ecosystem and we use whatever versions `grpc` needs."

@BillyONeal  - "That seems to be the thing that people want." 

@BillyONeal - "We should delist `upb`. https://github.com/protocolbuffers/protobuf/tree/main/upb ->  "For this reason, upb is not generally offered as a C library for direct consumption, and there are no releases."


Actions:
- Delist `upb` and instruct `grpc` to use its vendored copy
- Use the actual change that `srpc` accepted or even better update `srpc` instead of patching.

Questions:
1. Is there a newer version of `grpc` that still uses the 4.x series of `protobuf`? See, https://protobuf.dev/support/version-support/. If so, we could do an update to that while we wait for other ports to catch up.

Since upstream has been notified approx 3 ago, we will give them another 3 weeks to respond before we start delisting ports.

Tradias commented:
@JavierMatosD Great conversation. This matches my knowledge of grpc, protobuf, upb ecosystem exactly. I would throw utf8-range into the mix too but I think it is OK to keep it as is for now. I have been advocating for removing upb for a while now. Seeing that I have been doing the latest gRPC updates, I know how painful it is to patch CMake support into upb and patch gRPC to use it; two things that upstream never intended for.

GRPC v1.62.3 seems to be the last version to use protobuf 4.x: Look at the commit message on protobuf submodule between https://github.com/grpc/grpc/tree/v1.62.3/third_party with https://github.com/grpc/grpc/tree/v1.63.2/third_party

olejniczak commented:
Hi, it is great job, but the question is: when could it be merged? @JavierMatosD are all your requested changes already done?

Tradias committed:
openvino: Update protobuf patch to 2024.4

JavierMatosD commented:
@Tradias, please mark this PR "ready for review" when you want me to review this again.

Tradias commented:
@JavierMatosD what is the verdict regarding `upb` now? Are we merging this pr first and then delist it?

korDen commented:
Perhaps mark the PR as ready for review, otherwise it may not get the attention.

JavierMatosD commented:
@Tradias, sorry for the late response. I believe the action items requested by the team last we spoke were:

1. Delist `upb` and instruct `grpc` to use its vendored copy
2. Use the accepted change in `srpc` or update it to include the desired changes. 

We also agreed to give upstream 3 weeks to respond before we begin delisting ports. I believe it's been 3 weeks since `libarcus`, `brpc`, and `srpc` were notified.

Tradias committed:
grpc: No longer use upb port. Also use official protobuf-v5 patch for srpc

Tradias commented:
GRPC now uses built-in upb, tested successfully on x64-windows-static. Note that some patching is still needed to make it use the vcpkg utf8-range port.

SRPC now uses official patch, build successfully on x64-windows-static.

BRPC is also working on protobuf v5 support but I am waiting for it to be merged: https://github.com/apache/brpc/pull/2782

korDen commented:
Doesn't build for me on Microsoft (R) C/C++ Optimizing Compiler Version 19.38.33141 for x64

C:\develop\vcpkg\buildtrees\grpc\src\v1.65.5-3afb8dd835.clean\third_party\upb\upb\message\internal\message.c(22): error C2099: initializer is not a constant

JavierMatosD commented:
> GRPC now uses built-in upb, tested successfully on x64-windows-static. Note that some patching is still needed to make it use the vcpkg utf8-range port.
> 
> SRPC now uses official patch, build successfully on x64-windows-static.
> 
> BRPC is also working on protobuf v5 support but I am waiting for it to be merged: [apache/brpc#2782](https://github.com/apache/brpc/pull/2782)

It looks like https://github.com/apache/brpc/pull/2782 was merged. I'm assuming the `brpc` port should be updated to reflect?

korDen commented:
There is another issue that I'm seeing after updating, I didn't have it with previous snapshot (1.65.4?)

debug check failure at:
vcpkg\buildtrees\grpc\src\v1.65.5-5d18b9dc1c.clean\src\core\lib\event_engine\windows\grpc_polled_fd_windows.cc:227

CHECK(*from_len <= recv_from_source_addr_len_); // 128 <= 16

in 1.65.5 from_len is coming from
vcpkg\buildtrees\c-ares\src\v1.34.1-8a13123db5.clean\src\lib\ares_conn.c:53

where it's defined as

    struct sockaddr_storage sa_storage;
    ares_socklen_t          salen = sizeof(sa_storage); // 128

while in 1.65.4 it's coming from 
vcpkg\buildtrees\c-ares\src\v1.33.0-7b5315fac5.clean\src\lib\ares_process.c:448

where it's defined as
fromlen = sizeof(from.sa4); // 16

and then passed down.

Though I think the check itself may be flawed:

CHECK(*from_len <= recv_from_source_addr_len_);
memcpy(from, &recv_from_source_addr_, recv_from_source_addr_len_);

We are copying `recv_from_source_addr_len_` bytes into a `from` buffer, so we need to make sure that the buffer has AT LEAST `recv_from_source_addr_len_` bytes large, so I feel like the check should be `*from_len  >= recv_from_source_addr_len_`. I.e. if `from_len` is 10 and `recv_from_source_addr_len_` the check will pass but we will corrupt memory.

Tradias committed:
grpc: Fix upb NaN. Update brpc to 1.11.0 and add official protobuf v5. Download official srpc protobuf v5 patch

Tradias commented:
@JavierMatosD 

https://github.com/apache/brpc/pull/2782 was merged but the latest brpc release appeared 3 days before. I updated brpc and added the protobuf v5 fix as a patch.

I do not know how to delist `upb`. I cannot find any information on how to do that in this repo or here https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide

I added the `upb` NaN patch to `grpc`. Upstream has merged the fix https://github.com/protocolbuffers/protobuf/pull/17471 but it is not part of any release yet.

I preemptively added a patch for https://github.com/grpc/grpc/issues/37969 changing `<=` to `>=` in the assertion as @korDen and people in the issue discussion suggest.

BillyONeal committed:
Deindex upb.

BillyONeal commented:
> I do not know how to delist `upb`. I cannot find any information on how to do that in this repo or here https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide

Done

BillyONeal committed:
Revert upb addition to the version database following deindexing.

VAllens commented:
It's been a very long time since this PR, any progress?
I'm sure we all need it badly, anything to help move this forward?

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

VAllens commented:
This is great !
Thank you all so much for your efforts and contributions !
It makes for a thriving community !

AlangabSistemi commented:
Thank a lot to everyone!
